Common/src/Protocol/KeepaliveMonitor.ts

```diff
@@ -0,0 +1,78 @@
+import * as Messages from '../Messages/signalling_messages';
+import * as MessageHelpers from '../Messages/message_helpers';
+import { SignallingProtocol } from './SignallingProtocol';
+
+/**
+ * Used to regularly ping a protocol connection to make sure the connection is still good and open.
+ * When the pong doesn't come in response to a ping in time a callback is fired that can be handed
+ * by the owner.
+ */
+export class KeepaliveMonitor {
+    private protocol: SignallingProtocol;
+    private timeout: number;
+    private keepalive?: ReturnType<typeof setInterval>;
+    private alive: boolean = false;
+    private rtt: number = 0;
+
+    private onResponse: (pongMsg: Messages.pong) => void;
+
+    /**
+     * Called when a pong does not come back from a ping.
+     */
+    onTimeout?: () => void;
+
+    /**
+     * Gets the Round Trip Time of the current connection in milliseconds.
+     */
+    get RTT(): number {
+        return this.rtt;
+    }
+
+    /**
+     * Creates a new monitor and starts the ping timer. If a pong does not come back by the time we want
+     * to send a second ping then the connection is considered dead and the onTimeout callback is fired.
+     * @param protocol The connection that we want to monitor.
+     * @param timeout The time in milliseconds between ping messages.
+     */
+    constructor(protocol: SignallingProtocol, timeout: number) {
+        this.protocol = protocol;
+        this.timeout = timeout;
+        this.onResponse = this.onHeartbeatResponse.bind(this);
+        this.protocol.transport.on('close', this.stop.bind(this));
+        this.start();
+    }
+
+    private start(): void {
+        this.alive = true;
+        this.protocol.on('pong', this.onResponse);
+        this.keepalive = setInterval(this.sendHeartbeat.bind(this), this.timeout);
+    }
+
+    private stop(): void {
+        clearInterval(this.keepalive);
+        this.protocol.off('pong', this.onResponse);
```

Comment:
```suggestion
        this.protocol.off('pong', ()=>{ this.onHeartbeatResponse });
```

Seems like `onResponse` is always `onHeartbeatResponse` ?

Reply:
I did it this way so we can `on` and `off` the callback which I can't do if i use bind or an unnamed lambda.